File: nodejs/base/api/lib/routing/defaults/default-route-builder.ts
import { McmaApiRoute } from "../route";
import { McmaApiRouteHandler } from "../route-handler";
import { DefaultRouteHandlerConfigurator } from "./default-route-handler-configurator";
import { RouteStartedHandler } from "./route-started-handler";
import { RouteCompletedHandler } from "./route-completed-handler";
import { DefaultRouteBuilderType } from "./default-route-builder-type";

export class DefaultRouteBuilder<T> implements DefaultRouteBuilderType {
    private handler: McmaApiRouteHandler = null;
    
    constructor(private httpMethod: string, private path: string, private defaultRouteHandlerConfigurator: DefaultRouteHandlerConfigurator<T>) {}

    overrideHandler(handler: McmaApiRouteHandler): this {
        this.handler = handler;
        return this;
    }

    onStarted(handleOnStarted: RouteStartedHandler): this {
        this.defaultRouteHandlerConfigurator.onStarted = handleOnStarted;
        return this;
    }

    onCompleted(handleOnCompleted: RouteCompletedHandler<T>): this {
        this.defaultRouteHandlerConfigurator.onCompleted = handleOnCompleted;
        return this;
    }

    build(): McmaApiRoute {
        return new McmaApiRoute(this.httpMethod, this.path, this.handler ?? this.defaultRouteHandlerConfigurator.create());
    }
}

File: dotnet/base/Mcma.Api/Routing/Defaults/DefaultRouteBuilder.cs
using System.Threading.Tasks;
using System;
using System.Net.Http;
using Mcma.Api.Routes;

namespace Mcma.Api.Routing.Defaults
{
    public class DefaultRouteBuilder<TResult> : IDefaultRouteBuilder
    {
        internal DefaultRouteBuilder(HttpMethod method, string path, DefaultRouteHandlerConfigurator<TResult> defaultHandlerConfigurator)
        {
            Method = method;
            Path = path;
            DefaultRouteHandlerConfigurator = defaultHandlerConfigurator;
        }

        private HttpMethod Method { get; }

        private string Path { get; }

        private DefaultRouteHandlerConfigurator<TResult> DefaultRouteHandlerConfigurator { get; }

        internal Func<McmaApiRequestContext, Task> Handler { get; set; }

        public DefaultRouteBuilder<TResult> OverrideHandler(Func<McmaApiRequestContext, Task> handler)
        {
            Handler = handler;
            return this;
        }

        public DefaultRouteBuilder<TResult> OnStarted(Func<McmaApiRequestContext, Task<bool>> onStarted)
        {
            DefaultRouteHandlerConfigurator.OnStarted = onStarted; 
            return this;
        }

        public DefaultRouteBuilder<TResult> OnCompleted(Func<McmaApiRequestContext, TResult, Task> onCompleted)
        {
            DefaultRouteHandlerConfigurator.OnCompleted = onCompleted; 
            return this;
        }

        McmaApiRoute IDefaultRouteBuilder.Build() => Build();

        internal McmaApiRoute Build() => new McmaApiRoute(Method, Path, Handler ?? DefaultRouteHandlerConfigurator.Create());
    }
}
